File: dotnet/Operon.Sdk/Auth/DecodedClaims.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text.Json;
using Operon.Sdk.Internal;

namespace Operon.Sdk.Auth;

/// <summary>
/// Parsed JWT claims used by SDK components.
/// </summary>
public sealed class DecodedClaims
{
    public string? ParticipantDid { get; init; }
    public string? WorkstreamId { get; init; }
    public string? CustomerId { get; init; }
    public string? WorkspaceId { get; init; }
    public string? Email { get; init; }
    public string? Name { get; init; }
    public IReadOnlyList<string> TenantIds { get; init; } = Array.Empty<string>();
    public IReadOnlyList<string> Roles { get; init; } = Array.Empty<string>();
    public string? MemberId { get; init; }
    public string? SessionId { get; init; }
    public string? OrgId { get; init; }
    public string? ParticipantId { get; init; }
    public string? ClientId { get; init; }
    public string? AuthorizedParty { get; init; }
    public long ExpiresAtUnix { get; init; }

    public static DecodedClaims Decode(string token)
    {
        if (string.IsNullOrWhiteSpace(token))
        {
            return new DecodedClaims();
        }

        var segments = token.Split('.');
        if (segments.Length < 2)
        {
            return new DecodedClaims();
        }

        try
        {
            var payloadBytes = SdkModelHelpers.DecodeBase64Url(segments[1]);
            using var document = JsonDocument.Parse(payloadBytes);
            var root = document.RootElement;

            var workstream = ReadString(root, "workstream_id");
            if (string.IsNullOrWhiteSpace(workstream))
            {
                workstream = ReadString(root, "channel_id");
            }

            return new DecodedClaims
            {
                ParticipantDid = ReadString(root, "participant_did"),
                WorkstreamId = workstream,
                CustomerId = ReadString(root, "customer_id"),
                WorkspaceId = ReadString(root, "workspace_id"),
                Email = ReadString(root, "email"),
                Name = ReadString(root, "name"),
                TenantIds = ReadStringArray(root, "tenant_ids"),
                Roles = ReadStringArray(root, "roles"),
                MemberId = ReadString(root, "member_id"),
                SessionId = ReadString(root, "session_id"),
                OrgId = ReadString(root, "org_id"),
                ParticipantId = ReadString(root, "participant_id"),
                ClientId = ReadString(root, "client_id"),
                AuthorizedParty = ReadString(root, "azp"),
                ExpiresAtUnix = ReadLong(root, "exp")
            };
        }
        catch
        {
            return new DecodedClaims();
        }
    }

    private static string? ReadString(JsonElement element, string name)
    {
        if (element.TryGetProperty(name, out var value) && value.ValueKind == JsonValueKind.String)
        {
            var text = value.GetString();
            return string.IsNullOrWhiteSpace(text) ? null : text.Trim();
        }

        return null;
    }

    private static long ReadLong(JsonElement element, string name)
    {
        if (element.TryGetProperty(name, out var value) && value.ValueKind is JsonValueKind.Number)
        {
            return value.GetInt64();
        }

        return 0;
    }

    private static IReadOnlyList<string> ReadStringArray(JsonElement element, string name)
    {
        if (!element.TryGetProperty(name, out var value) || value.ValueKind != JsonValueKind.Array)
        {
            return Array.Empty<string>();
        }

        var list = new List<string>();
        foreach (var item in value.EnumerateArray())
        {
            if (item.ValueKind != JsonValueKind.String)
            {
                continue;
            }

            var text = item.GetString();
            if (!string.IsNullOrWhiteSpace(text))
            {
                list.Add(text.Trim());
            }
        }

        return list.ToArray();
    }
}

File: java/src/main/java/cloud/operon/sdk/auth/DecodedClaims.java
package cloud.operon.sdk.auth;

import com.fasterxml.jackson.databind.JsonNode;
import cloud.operon.sdk.internal.Json;

import java.io.IOException;
import java.util.ArrayList;
import java.util.Base64;
import java.util.Collections;
import java.util.List;

/**
 * Decoded JWT claims used across the SDK.
 */
public record DecodedClaims(
    String participantDid,
    String workstreamId,
    String customerId,
    String workspaceId,
    String email,
    String name,
    List<String> tenantIds,
    List<String> roles,
    String memberId,
    String sessionId,
    String orgId,
    String participantId,
    String clientId,
    String authorizedParty,
    long expiresAtUnix
) {

    public static DecodedClaims decode(String token) {
        try {
            String[] parts = token.split("\\.");
            if (parts.length < 2) {
                return empty();
            }

            byte[] payload = decodeBase64(parts[1]);
            JsonNode node = Json.mapper().readTree(payload);

            String workstream = text(node, "workstream_id");
            if (workstream == null || workstream.isBlank()) {
                workstream = text(node, "channel_id");
            }

            return new DecodedClaims(
                text(node, "participant_did"),
                workstream,
                text(node, "customer_id"),
                text(node, "workspace_id"),
                text(node, "email"),
                text(node, "name"),
                readArray(node, "tenant_ids"),
                readArray(node, "roles"),
                text(node, "member_id"),
                text(node, "session_id"),
                text(node, "org_id"),
                text(node, "participant_id"),
                text(node, "client_id"),
                text(node, "azp"),
                node.path("exp").isNumber() ? node.path("exp").asLong(0L) : 0L
            );
        } catch (IOException | IllegalArgumentException ex) {
            return empty();
        }
    }

    private static String text(JsonNode node, String field) {
        JsonNode value = node.path(field);
        if (value.isMissingNode() || value.isNull()) {
            return null;
        }
        String text = value.asText();
        return text == null || text.isBlank() ? null : text;
    }

    private static List<String> readArray(JsonNode node, String field) {
        JsonNode arr = node.path(field);
        if (!arr.isArray()) {
            return Collections.emptyList();
        }
        List<String> values = new ArrayList<>();
        arr.forEach(item -> {
            if (item.isTextual()) {
                String value = item.asText();
                if (!value.isBlank()) {
                    values.add(value);
                }
            }
        });
        return Collections.unmodifiableList(values);
    }

    private static byte[] decodeBase64(String value) {
        try {
            return Base64.getUrlDecoder().decode(value);
        } catch (IllegalArgumentException ex) {
            return Base64.getDecoder().decode(value);
        }
    }

    private static DecodedClaims empty() {
        return new DecodedClaims(
            null,
            null,
            null,
            null,
            null,
            null,
            Collections.emptyList(),
            Collections.emptyList(),
            null,
            null,
            null,
            null,
            null,
            null,
            0L
        );
    }
}
